src/python/pants/reporting/json_reporter.py

```diff
@@ -0,0 +1,145 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import json
+import os
+from builtins import list, open, str
+from collections import defaultdict, namedtuple
+
+from future.moves.itertools import zip_longest
+from six import string_types
+
+from pants.base.build_environment import get_buildroot
+from pants.reporting.reporter import Reporter
+from pants.util.dirutil import safe_mkdir
+
+
+class JsonReporter(Reporter):
+  """Json reporting to files."""
+
+  _log_level_str = [
+    'FATAL',
+    'ERROR',
+    'WARN',
+    'INFO',
+    'DEBUG',
+  ]
+
+  # JSON reporting settings.
+  #   json_dir: Where the report files go.
+  Settings = namedtuple('Settings', Reporter.Settings._fields + ('json_dir',))
+
+  def __init__(self, run_tracker, settings):
+    super(JsonReporter, self).__init__(run_tracker, settings)
+    # The main report, and associated tool outputs, go under this dir.
+    self._json_dir = settings.json_dir
+
+    self._buildroot = get_buildroot()
+    self._report_file = None
+    self._output_files = defaultdict(dict)
+
+    # Results of the state of the build
+    self._results = {}
+    # Workunit stacks partitioned by thread
+    self._stack_per_thread = defaultdict(list)
```

Comment:
I think `_stack_per_workunit_root` would be a more accurate name here. Workunits may have parents that start in a different thread.